### PR TITLE
New references model

### PR DESCRIPTION
**Pull request by rhaschke**
This is an attempt to simplify accessibility to references, making the references model essentially flat.
I also thought about distinguishing between _used_ and _available_ refs, such that the default list becomes much shorter...

# TODOs
- [x] Fix strange behavior of `RefComboBox` (always deleting the text in the lineedit)
- [x] Fix completer of `RefComboBox`: model with _all_ frames is correctly set, but not used. Works in `dev.py`!?!
- [ ] Allow insertion of new frames into `RefFramesModel` via GUI
   - [x] parent-property combo boxes of `Offset` frames allow insertion of (existing) `JointFrames`
   - [ ] reference-property combo boxes of tasks allow insertion of `JointFrames` or an arbitrarily named `Offset`.
   - [ ] I suggest temporarily creating a new interactive marker when editing an `Offset` frame.
   Optionally, the user could tick a checkbox to have that marker visible all the time. If so, the `Offset` frame should remember that fact. The GUI will need access to the `MarkerServer` to create (and delete) those markers on the fly.
- [x] ~~Add [`QValidator`](https://doc.qt.io/qt-5/qvalidator.html) to validate allowed frame names in a `RefComboBox`?~~
  The validator wasn't helpful for insertion of new frames.
- [ ] Update `RefFramesModel` on changes of `TaskHierarchy` (insert and remove frames) @felixkol
  Improve [`update_refs_from_tasks()`](https://github.com/ubi-agni/stack_of_tasks/blob/709734d304774f87336705dc9caacd377d838508/python/stack_of_tasks/ui/mainwindow.py#L52) and connect it to appropriate signals
- [ ] Use a `name` property in `RefFrame` and `Offset`?
     This would simplify the `RefFramesModel._refs` to a list of refs (instead of `List[Tuple[name, ref]]`)
- [ ] Drop `OffsetWithJacobian` and propagate `J` directly from the parent frame (if that property exists)?
  This is strongly related to #15
- [ ] `RelativeTasks` should offer an `offset` property too: the offset between the two reference frames. The corresponding marker should be created by a factory method of the task (a `PlaneTask` needs a different marker than a `PositionTask`). Maybe, only this marker should offer to stay visible all time.
   This `offset` provides the perfect setting to define e.g. a box constraint w.r.t. to the _configurable_ `A` frame, such that the alignment of the box can be adjusted to `A`. (If one wants to align the box with frame `B`, the user should simply switch `A` and `B`).
- [ ] What about a new `MarkerFrame` type inheriting from `Offset`? Such a marker would be automatically visible all time and could be reused by different reference frames.

# Documentation
We distinguish these types of reference frames:
- `RefFrame`: A frame that is inherently present in the ecosystem. Their names are _fixed_.
  - `World` (better `RootFrame`?): The root frame of the robot model. This is essentially a _constant_ identity transform.
  - `JointFrame` (better `LinkFrame`?): A robot link frame of the robot model.
  - `TfFrame`: A frame published by ROS TF
     Obviously, all robot link frames are published as TF frames as well (with the same names). Here, we prefer the corresponding `JointFrame`, whose transform can be updated directly via the robot's state.
- `Offset`: An arbitrarily named transform with an offset to a parent reference frame. The offset can be changed online, for example via an interactive marker. It is possible to nest `Offset` frames hierarchically.

### CODE REVIEW

**Comment by felixkol** [comment]
I really love the icons. 
The removal of TreeView in the ComboBox is probably the best choice, but I would like to keep it in overview, if possible.

File: python/stack_of_tasks/ui/model/av_ref.py [hunk 1]
```diff
@@ -11,15 +12,20 @@
 
 from . import RawDataRole
 
+icon_path = Path(rospkg.RosPack().get_path("rviz")) / "icons/classes"
+
+
+def get_ipath(icon: str, ext="png") -> str:
+    return str(icon_path / f"{icon}.{ext}")
+
```
Comment (rhaschke) [inline python/stack_of_tasks/ui/model/av_ref.py:20]: Isn't that overkill? Doing `QIcon(icon_path / "Axes.png")` below seems clearer (and not longer) to me.

File: python/stack_of_tasks/ui/model/av_ref.py [hunk 2]
```diff
@@ -11,15 +12,20 @@
 
 from . import RawDataRole
 
+icon_path = Path(rospkg.RosPack().get_path("rviz")) / "icons/classes"
```
Comment (rhaschke) [inline python/stack_of_tasks/ui/model/av_ref.py:15]: Don't pollute the namespace. Keep `icon_path` local to `AvailableRefModel.__init__()`.
Is the path slash (`/`) in `"icons/classes"` correctly resolved (i.e. on Windows using a different separator)?

**Comment by felixkol** [comment]
Sounds good to me. So the initial list of the joints-frames would be removed, joint frames can be added via the dialog. Then, we need to check, if the frame already exists on creation. Seems less of a mess compared to now.

**Comment by rhaschke** [comment]
Hi @felixkol,
I continued implementing the editable combobox for RefFrames. This was a nightmare because the combobox didn't behave as expected... The long list of frame names is correctly passed to the completer, but not used :cry:
Actually, sometimes it _was_ used, but not reproducible.

The `RefFramesModel` is now filled dynamically (currently only once during initialization):
https://github.com/ubi-agni/stack_of_tasks/blob/a1e5746e1f140faab015196297c9f94424b914dd/scripts/run_ui.py#L60-L63

To do this on the fly, you need to connect the corresponding signals...
Finally, we need to allow inserting new RefFrames from scratch.

**Pull request closed without merging**